xbmc/pvr/PVRDatabase.cpp

```diff
@@ -140,50 +140,45 @@ void CPVRDatabase::UpdateTables(int iVersion)
     VECADDONS addons;
     int iAddonId;
     CAddonDatabase database;
-    if (!database.Open() || !CAddonMgr::Get().GetAddons(ADDON_PVRDLL, addons, true))
-      return;
-
-    /** find all old client IDs */
-    std::string strQuery(PrepareSQL("SELECT idClient, sUid FROM clients"));
-    m_pDS->query(strQuery);
-    while (!m_pDS->eof() && !addons.empty())
+    if (database.Open() && CAddonMgr::Get().GetAddons(ADDON_PVRDLL, addons, true))
     {
-      /** try to find an add-on that matches the sUid */
-      iAddonId = -1;
-      for (VECADDONS::iterator it = addons.begin(); iAddonId <= 0 && it != addons.end();)
+      m_pDS->exec("ALTER TABLE channels ADD bClientIdUpgraded bool DEFAULT 0");
+
+      /** find all old client IDs */
+      std::string strQuery(PrepareSQL("SELECT idClient, sUid FROM clients"));
+      m_pDS->query(strQuery);
+      while (!m_pDS->eof() && !addons.empty())
       {
-        if ((*it)->ID() == m_pDS->fv(1).get_asString())
+        /** try to find an add-on that matches the sUid */
+        iAddonId = -1;
+        for (VECADDONS::iterator it = addons.begin(); it != addons.end(); ++it)
         {
-          /** try to get the current ID from the database */
-          iAddonId = database.GetAddonId(*it);
-          /** register a new id if it didn't exist */
-          if (iAddonId <= 0)
-            iAddonId = database.AddAddon(*it, 0);
-          if (iAddonId > 0)
+          if ((*it)->ID() == m_pDS->fv(1).get_asString())
           {
-            // this fails when an id becomes the id of one that's being replaced next iteration
-            // but since almost everyone only has 1 add-on enabled...
-            /** update the iClientId in the channels table */
-            strQuery = PrepareSQL("UPDATE channels SET iClientId = %u WHERE iClientId = %u", iAddonId, m_pDS->fv(0).get_asInt());
-            m_pDS->exec(strQuery);
-
-            /** no need to check this add-on again */
-            it = addons.erase(it);
-          }
-          else
-          {
-            ++it;
+            /** try to get the current ID from the database */
+            iAddonId = database.GetAddonId(*it);
+            /** register a new id if it didn't exist */
+            if (iAddonId <= 0)
+              iAddonId = database.AddAddon(*it, 0);
+            if (iAddonId > 0)
+            {
+              /** update the iClientId in the channels table */
+              strQuery = PrepareSQL("UPDATE channels SET iClientId = %u, bClientIdUpgraded = 1 WHERE iClientId = %u AND bClientIdUpgraded = 0", iAddonId, m_pDS->fv(0).get_asInt());
+              m_pDS->exec(strQuery);
+
+              /** no need to check this add-on again */
+              addons.erase(it);
+              break;
```

Comment:
if so then your break condition `iAddonId <= 0` in the for statement is wrong as well.
Apart from this I agree that we need to run through the complete list if it contains more than one entry (different version) per addon.